File: src/apple/Packages/MapleCore/Sources/MapleCore/Perspective.swift
// Perspective.swift — manual-geometry value type for `AdjustmentModel` (#3410).
//
// The seven `crs:Perspective*` scalars raw-core composes into one homography
// (`raw_core::stages::perspective`). raw-core declares them FLAT, as seven
// entries on `ADJUSTMENT_SCHEMA`; Swift groups them into this nested type the
// way it already groups the crop rect, because `AdjustmentModel.swift` cannot
// absorb seven more stored properties plus their memberwise-init parameters
// and assignments without breaching CONTRIBUTING.md's file budget — and its
// explicit `public init` is load-bearing (see that file's header), so the
// init cannot move to an extension to make room.
//
// The grouping is a Swift-side shape only. The generated `FieldName` cases
// stay flat and canonical (`perspectiveVertical`, …), the XMP keys are
// Adobe's, and `PresetAdjustmentBridge` maps each field name to
// `\.perspective.<member>` — so nothing outside this module can tell the
// difference.

import Foundation

/// Manual geometry: vertical and horizontal keystone, rotation, scale, aspect
/// stretch and X/Y offset. Mirror of `raw_core::stages::perspective::Perspective`.
///
/// Every member is normalised to the frame's own half-extents rather than to
/// pixels, so the same numbers describe the same correction on an image of any
/// size — which is what lets the live canvas, a downsampled preview and the
/// export master agree. The XMP serializer omits each key independently when it
/// still holds the value below.
public struct Perspective: Codable, Sendable, Equatable, Hashable {
    /// Vertical keystone, −100…100. Positive converges the bottom edge — the
    /// correction for a camera tilted up at a building. XMP
    /// `crs:PerspectiveVertical`.
    public var vertical: Double
    /// Horizontal keystone, −100…100. Positive converges the right edge. XMP
    /// `crs:PerspectiveHorizontal`.
    public var horizontal: Double
    /// Rotation in degrees, −10…10, positive = clockwise. Composes with
    /// `Crop.angle`'s ±45° straighten rather than replacing it: this one turns
    /// the frame the crop then samples. XMP `crs:PerspectiveRotate`.
    public var rotate: Double
    /// Uniform scale about the image centre, 50…150 percent. Above 100 pushes
    /// a keystone's transparent surround off-frame. XMP `crs:PerspectiveScale`.
    public var scale: Double
    /// Area-preserving aspect stretch, −100…100. Positive widens horizontally
    /// and compresses vertically by the reciprocal. XMP `crs:PerspectiveAspect`.
    public var aspect: Double
    /// Horizontal offset, −100…100, in hundredths of a half-extent — ±100 is
    /// half the frame width. XMP `crs:PerspectiveX`.
    public var x: Double
    /// Vertical offset, −100…100. XMP `crs:PerspectiveY`.
    public var y: Double

    public init(
        vertical: Double = 0,
        horizontal: Double = 0,
        rotate: Double = 0,
        scale: Double = 100,
        aspect: Double = 0,
        x: Double = 0,
        y: Double = 0
    ) {
        self.vertical = vertical
        self.horizontal = horizontal
        self.rotate = rotate
        self.scale = scale
        self.aspect = aspect
        self.x = x
        self.y = y
    }

    /// No manual geometry — every member at the value that makes its factor
    /// the identity matrix.
    public static let identity = Perspective()

    /// True when the composed homography is exactly the identity, which is
    /// what both render paths short-circuit on.
    public var isIdentity: Bool {
        self == .identity
    }
}

File: src/raw-pipeline/raw-core/src/types/adjustment/schema/perspective.rs
//! Manual-geometry schema entries (#3410) — Adobe's `crs:Perspective*` keys,
//! split out of `schema/mod.rs` to keep that file under the 570-line headroom
//! budget (CONTRIBUTING.md). `ADJUSTMENT_SCHEMA` lists these consts in place,
//! so schema order still matches struct order.
//!
//! Seven scalars composing into ONE homography (`stages::perspective`),
//! applied between EXIF orientation and the user crop. Ranges and defaults
//! are Adobe's, so a Lightroom sidecar's values load without rescaling; the
//! normalized coefficients each unit maps to are pinned by the constants in
//! `stages::perspective::matrix`.

use super::{FieldKind, FieldSpec};

/// A manual-geometry slider: `F32`, range −100..100, default 0. Five of the
/// seven share that shape exactly; `rotate` and `scale` spell their own out.
const fn geometry_field(name: &'static str, doc: &'static str) -> FieldSpec {
    FieldSpec {
        name,
        kind: FieldKind::F32,
        range: (-100.0, 100.0),
        default_f32: 0.0,
        enum_name: "",
        doc,
    }
}

pub(super) const PERSPECTIVE_VERTICAL: FieldSpec = geometry_field(
    "perspective_vertical",
    "Vertical keystone correction (#3410, `crs:PerspectiveVertical`). Projective coefficient about the image centre: positive converges the bottom edge (the correction for a camera tilted up at a building). 0 (default) contributes an identity row to the homography.",
);

pub(super) const PERSPECTIVE_HORIZONTAL: FieldSpec = geometry_field(
    "perspective_horizontal",
    "Horizontal keystone correction (#3410, `crs:PerspectiveHorizontal`). Projective coefficient about the image centre: positive converges the right edge. 0 (default) is identity.",
);

pub(super) const PERSPECTIVE_ROTATE: FieldSpec = FieldSpec {
    name: "perspective_rotate",
    kind: FieldKind::F32,
    range: (-10.0, 10.0),
    default_f32: 0.0,
    enum_name: "",
    doc: "Geometry rotation in degrees, positive = clockwise (#3410, `crs:PerspectiveRotate`). Adobe's ±10° fine level, independent of `crop.angle`'s ±45° straighten: this one rotates INSIDE the frame the crop then samples, so the two compose rather than replace one another. 0 (default) is identity.",
};

pub(super) const PERSPECTIVE_SCALE: FieldSpec = FieldSpec {
    name: "perspective_scale",
    kind: FieldKind::F32,
    range: (50.0, 150.0),
    default_f32: 100.0,
    enum_name: "",
    doc: "Uniform scale about the image centre, in percent (#3410, `crs:PerspectiveScale`). Below 100 shrinks the frame's content inward (exposing the transparent surround a keystone leaves behind); above 100 magnifies it to push that surround off-frame. 100 (default) is identity.",
};

pub(super) const PERSPECTIVE_ASPECT: FieldSpec = geometry_field(
    "perspective_aspect",
    "Aspect stretch (#3410, `crs:PerspectiveAspect`). Positive stretches horizontally and compresses vertically by the reciprocal factor, so frame area is preserved; negative does the opposite. 0 (default) is identity.",
);

pub(super) const PERSPECTIVE_X: FieldSpec = geometry_field(
    "perspective_x",
    "Horizontal offset of the transformed frame (#3410, `crs:PerspectiveX`). ±100 shifts by one half-extent — half the frame width. 0 (default) is identity.",
);

pub(super) const PERSPECTIVE_Y: FieldSpec = geometry_field(
    "perspective_y",
    "Vertical offset of the transformed frame (#3410, `crs:PerspectiveY`). ±100 shifts by one half-extent — half the frame height. 0 (default) is identity.",
);
